### ISSUE TITLE
IntegrityError: new row for relation "sounds_sound" violates check constraint "sounds_sound_num_comments_check"

### ISSUE THREAD

**Issue by ffont** [MTG/freesound#894]
https://logserver.mtg.upf.edu/sentry/freesound-web/issues/666/

```
IntegrityError: new row for relation "sounds_sound" violates check constraint "sounds_sound_num_comments_check"
DETAIL:  Failing row contains (393776, 3481306, 2017-05-23 11:00:10.914712+02, NPR Humpty or Ned.mp3, /home/fsweb/freesound/freesound-data/sounds/393/393776_3481306.m..., 393776__breeze87__npr-humpty-or-ned, humpy dumpy is going through a lot of changes, one of them is hi..., null, 2, null, 27528, mp3, 60.2405, 345, 0, 22050, 305546, 1, f1acabdcfa9ca7e09dd495cd5404e5df, t, OK, 2017-06-16 10:57:42.657304+02, null, f, OK, 2017-05-23 11:00:16.455642+02, null, OK, OK, -1, 0, 0, 0, f7f04eed, null, FI, f).

(23 additional frame(s) were not displayed)
...
  File "comments/models.py", line 48, in on_delete_comment
    instance.content_object.post_delete_comment()
  File "sounds/models.py", line 625, in post_delete_comment
    self.save()

IntegrityError: new row for relation "sounds_sound" violates check constraint "sounds_sound_num_comments_check"
DETAIL:  Failing row contains (393776, 3481306, 2017-05-23 11:00:10.914712+02, NPR Humpty or Ned.mp3, /home/fsweb/freesound/freesound-data/sounds/393/393776_3481306.m..., 393776__breeze87__npr-humpty-or-ned, humpy dumpy is going through a lot of changes, one of them is hi..., null, 2, null, 27528, mp3, 60.2405, 345, 0, 22050, 305546, 1, f1acabdcfa9ca7e09dd495cd5404e5df, t, OK, 2017-06-16 10:57:42.657304+02, null, f, OK, 2017-05-23 11:00:16.455642+02, null, OK, OK, -1, 0, 0, 0, f7f04eed, null, FI, f).

```

**Comment by ffont** [comment]
This happened to me when trying to delete a user account which had made comments to some sounds and the `num_comments` property of the `sound` object did not match `sound.comments.all().count()`

**Comment by alastair** [comment]
I had a look at the code around counting comments, and saw a handful of other things that I think should be addressed around the way that we deal with comments. While they're not strictly related to this specific error, I think it's good to leave our code in a better state than when we look at, even if the changes are not specifically related to our bug.

With that said, here's a list of things:

 * [Comment](https://github.com/MTG/freesound/blob/878a435a1e0a7921e6dd523c3dc09192ce8fed6a/comments/models.py#L34) uses `GenericForeignKey`s but we only ever have comments on sounds. For simplicity I would advocate that we remove this and just link them directly to `Sound`. In fact, [we already have some places which make this assumption](https://github.com/MTG/freesound/blob/878a435a1e0a7921e6dd523c3dc09192ce8fed6a/apiv2/views.py#L464).
 * [Sound.add_comment](https://github.com/MTG/freesound/blob/878a435a1e0a7921e6dd523c3dc09192ce8fed6a/sounds/models.py#L612) takes as an argument a Comment object which has not been saved. It then saves it, which is a weird separation of concerns. A simpler solution would be to pass in a user/comment text directly to this method and have the method make the Comment.
 * Sound.add_comment has an argument `commit=True` which isn't used. If it's not used, we shouldn't have it. It seems like it was added for consistency with other methods which save the Sound, but I think that we shouldn't have it.
 * [Sound.post_delete_comment](https://github.com/MTG/freesound/blob/878a435a1e0a7921e6dd523c3dc09192ce8fed6a/sounds/models.py#L621) has a confusing name. I didn't understand that it was called on the `post_delete` signal of a *Comment* until searching and following the code path. This could be improved by renaming the method or adding documentation to better learn what it does.

To fix the actual issue (number of comments becoming out of sync with the counter) we should investigate more deeply into what is causing the problem. @ffont has a script which checks for mismatching values, and we noticed one sound which had the same comment twice. We wonder if this issue may be happening when someone double-clicks on the submit button when making a comment (similar: #742, #818, #819).

I can believe that two subsequent requests get a `Sound` object and then both increment the counter by 1 and save the sound. The counter will be 1 more than it was, even though two comments have been made.

This could be fixed by using [F-expressions](https://docs.djangoproject.com/en/dev/ref/models/expressions/#f-expressions) or by performing the full Comment create POST in a single atomic transaction. The solution should be chosen keeping in mind the other places where we might want to fix similar issues.

**Comment by andrebola** [comment]
The bullets are implemented on PR #907 and also using the F-expressions for updating number of comments. Now I will continue to work on the related issues:  #742, #818, #819 but I will open new PR for that
